### DISCUSSION TITLE
After migration, I get a runtime error re Mongo/BSON `_id` serialization—this wasn't mentioned in the migration guide

### DISCUSSION THREAD

**Discussion by ghost** [sveltejs/kit#6194] [Q&A]
Here the error message:

```
data.car._id.toHexString returned from 'load' in src/routes/cars/[carId]/+page.server.ts cannot be serialized as JSON
Error: data.car._id.toHexString returned from 'load' in src/routes/cars/[carId]/+page.server.ts cannot be serialized as JSON
    at check_serializability (file:///carshop/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.429_svelte@3.49.0+vite@3.0.5/node_modules/@sveltejs/kit/src/runtime/server/page/load_data.js:134:8)
    at check_serializability (file:///carshop/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.429_svelte@3.49.0+vite@3.0.5/node_modules/@sveltejs/kit/src/runtime/server/page/load_data.js:128:5)
    at check_serializability (file:///carshop/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.429_svelte@3.49.0+vite@3.0.5/node_modules/@sveltejs/kit/src/runtime/server/page/load_data.js:128:5)
    at check_serializability (file:///carshop/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.429_svelte@3.49.0+vite@3.0.5/node_modules/@sveltejs/kit/src/runtime/server/page/load_data.js:128:5)
    at load_server_data (file:///carshop/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.429_svelte@3.49.0+vite@3.0.5/node_modules/@sveltejs/kit/src/runtime/server/page/load_data.js:34:3)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async file:///carshop/node_modules/.pnpm/@sveltejs+kit@1.0.0-next.429_svelte@3.49.0+vite@3.0.5/node_modules/@sveltejs/kit/src/runtime/server/page/index.js:163:13
```
and `_id` is a Mongo bson id, before all worked fine. 

Is this related to https://github.com/sveltejs/kit/discussions/6143 and https://github.com/sveltejs/kit/issues/6008 ?

What is the recommended migration way?

**Comment by infovore** [comment]
Hard to tell without seeing your actual `+page.server.ts` code - but yes, it's definitely related to the `load` function trying to return something that can't be serialized - from the looks of it, a function. (the function `toHexString` which is part of the id object. I don't know enough about Mongo/bson to be able to comment further without seeing the code, and that means it's hard to tell if #6008 would fix it, or if there's still a tweak needed.

**Reply by ghost** [reply]
> to be able to comment further without seeing the code

I just return a result, a document, from MongoDB, nothing more:
```typescript
...
if (result)
  return {
    car: result
  }
...
```
This result/document includes as a Mongo document always its id, a so called ObjectId with the key `_id` and more in case of foreign keys.

I assume I need now a serializer which supports ObjectId, e.g., superjson but how do I integrate it?

**Reply by infovore** [reply]
well - you could serialize it by hand, ie, rather than just returning the result, transforming it yourself, or running it through a serializer in `+page.server.ts` and then, on object load inside `+page.svelte`, unserializing. Not knowing enough about this particular Mongo object: do you need all the methods on it, or is the raw JSON representation enough? (The sticking point will probably be `Date`s if you have them).

Looking at the docs, that seems like it might be as straightforward as importing `superjson` and then using `superjson.serialize(result)`, and then `superjson.deserialize(car)` on the page component.

**But**: that's almost certainly not going to serialize `toHexString` because it's a _method_, not a _property_. I don't know if you even need the methods or not. But you're not going to be able to pass an object with functions from back-end to front-end, and I don't know how you were before - unless your queries were _only_ running on the client.

**Reply by infovore** [reply]
Honestly, as a starting point, try:

```
if(result)
  return {
    car: JSON.parse(JSON.serialize(result))
  }
}
```

which will send you back a pure JSON object representation of `result`, with methods stripped. That may be all you need. (It was on an issue I had related to this). If you were to say "that feels ugly", I'd agree, and there are likely better ways to do this. However, it may also reveal that you need "real" serialisation, owing to the usage of things like more complex fields.

**Comment by AlexRMU** [comment]
The fact is that _id is an object. You need to do [lean](https://mongoosejs.com/docs/tutorials/lean.html) or use [json](https://kit.svelte.dev/docs/modules#sveltejs-kit-json).
